Deprecate mutating a configuration after observation (#27558)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

        }
    }
}

tasks.register("resolve") {
    val conf = configurations["runtimeClasspath"]

    // Wire build dependencies
    dependsOn(conf)

    // Resolve dependencies
    doLast {
        assert(conf.files.map { it.name } == listOf("foo-1.0.jar"))
    }
}
----
======
=====

For the following cases, consider the suggested options for replacing mutations within a `beforeResolve` hook:

* **Adding dependencies**: Use link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#withDependencies-org.gradle.api.Action-[`withDependencies`].
* **Changing dependency versions**: Use dependency constraints. (See <<using_preferred_versions, Declaring preferred version constraints>>)
* **Adding excludes**: Use <<component_metadata_rules.adoc#sec:component_metadata_rules,Component Metadata Rules>>
* **Roles**: Configuration roles should be set upon creation and not changed afterward.
* **Hierarchy**: Configuration hierarchy (`extendsFrom`) should be set upon creation. Mutating the hierarchy prior to resolution is highly discouraged, but is permitted within a `withDependenices` hook.

> * **Resolution Strategy**: Mutating a configuration's ResolutionStrategy is still permitted in a `beforeResolve` hook, however this is not a recommended practice.

Reviewer: ```suggestion
* **Resolution Strategy**: Mutating a configuration's ResolutionStrategy is still permitted in a `beforeResolve` hook; however, this is not recommended.
```

=== Deprecations

[[mutate_configuration_after_locking]]
==== Deprecate mutating configuration after observation

To ensure the accuracy of dependency resolution, Gradle ensures that Configurations are not mutated after they have been used as part of a dependency graph.

> * Resolvable configurations should not have their resolution strategy, dependencies, hierarchy, etc., modified after they have been resolved.

Reviewer: FYI: etc., is correct.

> * Resolvable configurations should not have their resolution strategy, dependencies, hierarchy, etc., modified after they have been resolved.

Author: ChatGPT says: 
```
No, a comma after "etc." is not necessary in this context. The rule for using a comma after "etc." varies depending on the specific style guide you follow, but generally, in American English, when "etc." appears at the end of a list within a sentence, it is not followed by a comma. Your sentence is correctly punctuated as it stands.
```

* Consumable configurations should not have their dependencies, hierarchy, attributes, etc., modified after they have been published or consumed as a variant.
* Dependency scope configurations should not have their dependencies, constraints, etc., modified after a configuration that extends from them are observed.

> Gradle already throws an exception for many of these circumstances, but in some cases, it does not.

Author: ```suggestion
In prior versions of Gradle, many of these circumstances were detected and handled by failing the build. 
However, some cases went undetected or did not trigger build failures.
```

----
plugins {
    id("java-library")
}

configurations.runtimeClasspath {
    // `beforeResolve` is called after the configuration is resolved for task dependencies,
    // but before it is fully resolved during artifact resolution.
    // Configurations should not be mutated in this hook
    incoming.beforeResolve {
        if (allDependencies.none { it.group == "com" && it.name == "foo" }) {
            configurations.implementation.get().dependencies.add(project.dependencies.create("com:foo:1.0"))
        }
    }
}

tasks.register("resolve") {
    val conf = configurations["runtimeClasspath"]

    // Wire build dependencies
    dependsOn(conf)

    // Resolve dependencies
    doLast {
        assert(conf.files.map { it.name } == listOf("foo-1.0.jar"))
    }
}
----
======
=====

For the following cases, consider the suggested options for replacing mutations within a `beforeResolve` hook:

* **Adding dependencies**: Use link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#withDependencies-org.gradle.api.Action-[`withDependencies`].
* **Changing dependency versions**: Use dependency constraints. (See <<using_preferred_versions, Declaring preferred version constraints>>)

> * **Adding excludes**: Use <<component_metadata_rules.adoc#sec:component_metadata_rules,Component Metadata Rules>>

Reviewer 2: ```suggestion
* **Adding excludes**: Use <<component_metadata_rules.adoc#sec:component_metadata_rules,Component Metadata Rules>>.
```

> * **Roles**: Configuration roles should be set upon creation and not changed afterward.

Reviewer 2: Should this link to the new APIs?

Author: There was a release notes entry for these, but we don't really have the new APIs documented anywhere.

We could link to the release notes of _another_ version of Gradle but that seems weird (since `notes.md` is overwritten each Gradle version)

> * **Hierarchy**: Configuration hierarchy (`extendsFrom`) should be set upon creation. Mutating the hierarchy prior to resolution is highly discouraged, but is permitted within a `withDependenices` hook.

Reviewer 2: ```suggestion
* **Hierarchy**: Configuration hierarchy (`extendsFrom`) should be set upon creation. Mutating the hierarchy prior to resolution is highly discouraged, but is permitted within a `withDependencies` hook.
```
